articles/an-introduction-to-memoization-in-javascript/index.md

```diff
@@ -0,0 +1,381 @@
+  let result = 0;
+  for (let i = 1; i <= num; i++) {
+    for (let j = 1; j <= num; j++) {
+      result ++;
+    }
+  }
+  return result;
+}
+console.log(clumsysquare(4));
+console.log(clumsysquare(10));
+console.log(clumsysquare(12));
+console.log(clumsysquare(17));
+console.log(clumsysquare(20));
+```
+
+You will realize that it will re-execute the function whenever you call it and then return a squared value.
+
+The execution is straightforward and fast. The computation is straightforward. The real problem occurs when you want to perform heavy (expensive) computation.
+
+```js
+const clumsysquare = num =>{
+  let result = 0;
+  for (let i = 1; i <= num; i++) {
+    for (let j = 1; j <= num; j++) {
+      result ++;
+    }
+  }
+  return result;
+}
+console.log(clumsysquare(190));
+console.log(clumsysquare(799));
+console.log(clumsysquare(4000));
+console.log(clumsysquare(7467));
+console.log(clumsysquare(9666));
+```
+
+This is not efficient. Instead, we can apply the concept of Memoization to store the result of `clumsysquare()` when first called. Whenever we call this function with the same input, the program will not have to execute it again.
+
+Let's implement Memoization in our JavaScript program. To do that, the program executes the first instance `clumsysquare()` we’ll store its value then reuse it several times within the program.
+
+```js
+// function that takes a function and returns a function
+const memoize = (func) => {
+  // a cache of results
+  const results = {};
+  // return a function for the cache of results
+  return (...args) => {
+    // a JSON key to save the results cache
+    const argsKey = JSON.stringify(args);
+    // execute `func` only if there is no cached value of clumsysquare()
+    if (!results[argsKey]) {
+      // store the return value of clumsysquare()
+      results[argsKey] = func(...args);
+    }
+    // return the cached results
+    return results[argsKey];
+  };
+};
+
+// wrap clumsysquare() in memoize()
+const clumsysquare = memoize(num => {
+    let result = 0;
+    for (let i = 1; i <= num; i++) {
+        for (let j = 1; j <= num; j++) {
+            result++;
+        }
+    }
+    return result;
+});
+
+console.log(clumsysquare(190));
+console.log(clumsysquare(799));
+console.log(clumsysquare(4000));
+console.log(clumsysquare(7467));
+console.log(clumsysquare(9666));
+```
+
+The example above executes faster compared to when not using the concept of Memoization. You can choose to store the cache using another function like shown above. Alternatively, you chose to store the results in a variable. Check this example.
+
+```js
+const memoizedValue = [];
+
+const clumsysquare = (num) => {
+    if ((memoizedValue[num] = !null)) {
+        return memoizedValue[num];
+    }
+
+    let result = 0;
+    for (let i = 1; i <= num; i++) {
+        for (let j = 1; j <= num; j++) {
+            result++;
+        }
+    }
+
+    memoizedValue[num] = result;
+    return result;
+};
```

Comment:
@kimkimani This function is always returning `true` instead of the square value. 
It's probably because of the incorrect operator in the `if` statement. You should use `!==`.

Please make sure this function works properly.